### ISSUE TITLE
Cannot load package assets from a directory

### ISSUE THREAD

**Issue by kenzieschmoll** [flutter/flutter#112019]
Here is the package structure we are working with:
```
- devtools/
  - packages/
    - devtools_app/
      - pubspec.yaml
  - third_party/
    - packages/
      - perfetto_compiled/ (this folder contains html, css, js, etc. assets for the perfetto UI web app)
```

In the `devtools_app/pubspec.yaml` file:
```
flutter
  assets:
    packages/perfetto_compiled/
```

At this point, running `flutter build web` fails with the error:
```
Error: unable to find directory entry in pubspec.yaml: /Users/kenzieschmoll/develop/devtools/packages/devtools_app/packages/perfetto_compiled/
```

Specifying the full asset paths for files under `packages/perfetto_compiled/` builds successfully. I would have expected this to work based on the Flutter documentation around specifying assets (https://docs.flutter.dev/development/ui/assets-and-images#specifying-assets).

I was unable to confirm whether this is a web only issue - `flutter build macos` was failing for me with signing issues.

**Comment by jonahwilliams** [comment]
```
Error: unable to find directory entry in pubspec.yaml: /Users/kenzieschmoll/develop/devtools/packages/devtools_app/packages/perfetto_compiled/
```

This seems like a bogus message that we should fix.  IIRC the intention of packages/<package-name> is to pull assets from pub dependencies. Is `perfetto_compiled` something that is in pubspec.yaml?

Its possible that something goes wrong in a way we don't handle well if the packages path is used with something that isn't a dependency

**Comment by kenzieschmoll** [comment]
Yes `perfetto_compiled` is a dependency in the pub spec. I forgot to include that in the pubspec above - edited the original description to show it.

**Comment by andrewkolos** [comment]
I was able to reproduce on the master channel (8a21acc7d5). [Repro](https://github.com/andrewkolos/flutter_issue_demos_2/tree/2d7a723e28f67db51808d2579a3adf2c7175db92). I'll do some investigation.

**Comment by andrewkolos** [comment]
@kenzieschmoll Where are the image assets stored within the `perfetto_compiled` directory? If I'm reading the [Asset images in package dependencies](https://docs.flutter.dev/development/ui/assets-and-images#bundling-of-package-assets) section in the docs correctly, the way to include assets from other directories without specifying each file individually is to have the assets reside within the `lib` folder of the package you are depending on. When I did this, I was able to import an image without specifying the full path to it in the pubspec. I got this to work in the same [repro repo](https://github.com/andrewkolos/flutter_issue_demos_2/tree/main) I linked in my last message.

**Comment by kenzieschmoll** [comment]
The assets in `perfetto_compiled` are under `devtools/third_party/packages/perfetto_compiled/lib/dist/v29.0-3a9f96aec/assets/`

**Comment by andrewkolos** [comment]
> Specifying the full asset paths for files under packages/perfetto_compiled/ builds successfully.

Taking a look at the docs again, I think I was confused. I think what you described here is intended for bundling package assets from packages that do not specify the file in their own pubspec. Quoting the docs,

> If the desired asset is specified in the pubspec.yaml file of the package, it’s bundled automatically with the application. In particular, assets used by the package itself must be specified in its pubspec.yaml.
A package can also choose to have assets in its lib/ folder that are not specified in its pubspec.yaml file. **In this case, for those images to be bundled, the application has to specify which ones to include in its pubspec.yaml.** For instance, a package named fancy_backgrounds could have the following files:

I do agree that the false error message is interesting.

@jonahwilliams Please double check me on this.

**Comment by kenzieschmoll** [comment]
So if I were to specify the perfetto_compiled assets in `perfetto_compiled/pubspec.yaml` could I think include them in `devtools_app/pubspec.yaml` as such?
```
flutter
  assets:
    packages/perfetto_compiled/
```
It's worth noting that the `perfetto_compiled/` package is not a flutter app. Here is the `perfetto_compiled/pubspec.yaml`:

```
name: perfetto_compiled
description: Pre-compiled Perfetto UI web app with Flutter DevTools theme support.
homepage: https://github.com/flutter/devtools/tree/master/third_party/packages/perfetto_compiled
version: 0.0.1

environment:
  sdk: '>=2.12.0 <3.0.0'
```

**Comment by andrewkolos** [comment]
Sorry, I completely missed the notification for this.

> So if I were to specify the perfetto_compiled assets in `perfetto_compiled/pubspec.yaml` could I think include them in `devtools_app/pubspec.yaml` as such?
> 
> ```
> flutter
>   assets:
>     packages/perfetto_compiled/
> ```

Sure. Suppose the image you want to use is at `devtools/third_party/packages/perfetto_compiled/lib/dist/v29.0-3a9f96aec/assets/dog.jpg`, then

1. Add the folder to `perfetto_compiled/pubspec.yaml`:
```yaml
flutter:
  assets:
    - lib/dist/v29.0-3a9f96aec/assets/
```

2.  In `devtools_app`, wherever you want to show your image, you can load it like so:

```dart
Image(image: AssetImage('packages/perfetto_compiled/lib/dist/v29.0-3a9f96aec/assets/dog.jpeg'))
```

> It's worth noting that the `perfetto_compiled/` package is not a flutter app. Here is the `perfetto_compiled/pubspec.yaml`:
> 
> ```
> name: perfetto_compiled
> description: Pre-compiled Perfetto UI web app with Flutter DevTools theme support.
> homepage: https://github.com/flutter/devtools/tree/master/third_party/packages/perfetto_compiled
> version: 0.0.1
> 
> environment:
>   sdk: '>=2.12.0 <3.0.0'
> ```

This is fine, you can still add a `flutter` entry to the pubspec, even if it's not a flutter app itself.

____
If you don't want to modify the pubspec of `perfetto_compiled`, you can still bundle your assets another way since your assets are within the `lib` folder of `perfetto_compiled`:

1. In the `devtools_app` pubspec, specify the _full_ path to _each_ asset:
```yaml
flutter:
  assets:
    - packages/perfetto_compiled/dist/v29.0-3a9f96aec/assets/dog.jpeg
```
2. Load your image like so (notice that the `lib` folder in the path is now implicit):
```dart
Image(image: AssetImage('packages/perfetto_compiled/dist/v29.0-3a9f96aec/assets/dog.jpeg'))
```

**Comment by andrewkolos** [comment]
The pubspec files look fine, but how are you loading the image within `devtools_app`? Using this from my previous comment, I am able to get the image to load.

> 3\. In `devtools_app`, wherever you want to show your image, you can load it like so:
> 
> ```dart
> Image(image: AssetImage('packages/perfetto_compiled/lib/dist/v29.0-3a9f96aec/assets/dog.jpeg'))
> ```
> 
> >

Even though the error message you mentioned will appear, the image should still get loaded properly. As far as I can tell, that error message is erroneous, and I will get it documented as a new Issue once I invest time in better understanding what's causing it.

**Comment by andrewkolos** [comment]
Small amendment to my previous two comments. In the setup we are using, `demo_app` doesn't need to explicitly declare assets from `perfetto_compiled`, because `perfetto_compiled` already declares them.

That is, as long as you 
1) have `perfetto_compiled` declared as a dependency of `devtools_app` 
 ```yaml
# devtools_app pubspec.yaml
dependencies:
  flutter:
    sdk: flutter
  perfetto_compiled:
    path: ../../third_party/packages/perfetto_compiled
```
2) and have `perfetto_compiled` explicitly declare assets in its pubspec, `devtools_app` should be able to those assets.
```yaml
# perfetto_compiled pubspec.yaml
flutter:
  assets:
    - lib/dist/v29.0-3a9f96aec/assets/
```

The image should load just as before, and you shouldn't get any errors in the build output.

<details>
<summary>chrome demo</summary>

![image](https://user-images.githubusercontent.com/9027551/199099269-755c605a-a307-4e98-9aac-a9eb8bb3bd54.png)

</details>
